### ISSUE TITLE
[Bug]: Column content lost on data entry

### ISSUE THREAD

**Issue by bernard01** [sqlitebrowser/sqlitebrowser#3248]
### What did you do?

I created a simple database with an integer primary key and a few text columns, one of them not nullable. The primary key values are manually generated, so records can be entered out of order.

Then I entered a new record with key 1 in the table like a spreadsheet and saved it.
Then I entered a new record with key 0 as above and saved it.

Next I sorted the 2 rows by clicking on the primary key column header.

The attached zip file contains the database files:

[ColumnContentErased.zip](https://github.com/sqlitebrowser/sqlitebrowser/files/10288568/ColumnContentErased.zip)

### What did you expect to see?

I expect to see that there are 2 records in the database with values as entered.

### What did you see instead?

For an unknown reason, in browse mode, content of all columns except the integer primary key is erased for a new record. With erased I mean that nullable columns are null, and other columns that are not nullable contain empty strings.

This appears to happen on insert of new rows.

It does not become visible immediately but after sorting of the view.

I could not believe this at first because I have other databases on the same computer that work fine. So I checked the file system. The file system looks sound. I did a check disk and I made a full backup which ran without errors.

I tried it again with a new database. Same problem. I can work around the problem as follows:

I enter a new record with the primary key only and save it. Then I re-order it and back into the natural order. Which means that I have to search for my new record in the table which is easy because it is basically blank.

Then I update that record which avoids the data loss.

I suspect that the software is confused by the fact that I enter the rows out of primary key order, and that because of it, due to its initial default ordering, it loses my data entry even while I save the record. Still the data is shown because the GUI is stale at that point. But when I re-order the rows then I see that the data was not actually saved, only the primary key.

### DB4S Version

3.12.2

### What OS are you seeing the problem on?

Windows

### OS version

Windows 7

### Relevant log output

```shell
I do not know how to get any log output at all because I have never seen a console or log file.
```

### Prevention against duplicate issues

- [X] I have searched for similar issues

**Comment by chrisjlocke** [comment]
First off, thanks for reporting this, rather than banging the keyboard in frustration, kicking the cat and swearing at something .. as the latter is something I would have done.
Have a quick read of https://github.com/sqlitebrowser/sqlitebrowser/issues/3240 as that's a similar issue, and one that's bitten you too.
Again, it sort of makes sense, but DB4S could handle it better.

What I believe has happened is that you've edited the primary key, which has 'confused' the grid.  As per the issue linked to above, edit the primary key, /then/ do all the other updates.

Have another go, and let me know if this issue is the same.

You'll notice in the other issue, the person reporting didn't report back ... so please report back if this or is not the case so we can assist further.  We do try to help.

**Comment by bernard01** [comment]
Sorry, it's not the same problem. I have read 600 issues before I reported this. Could not find a duplicate. I definitely did not change the primary key. BTW I got the problem with fairly high primary key numbers so the 0 zero number as in the test case is not significant. The problem is pretty much as I described it. And I can reproduce it every time.

I know it's fairly unbelievable and I really had a bad day when I thought I could not trust this DB any more. It's so good otherwise. Then I worked my way through it with running the attached SQL query after every step I did and caught it. There is absolutely nothing special I am doing. Not even cut and paste. The tricky bit was that in real life, the negative effect is delayed because the saved corrupted record moves out of the field of view and only after a dozen or so entries I realized that I had a lot of nulls in the table.

**Comment by chrisjlocke** [comment]
OK, bear with me.  I've not seen this before - only the issue with editing the primary key.
Being reproduceable is a good thing.

> I enter a new record with the primary key only and save it. Then I re-order it and back into the natural order. Which means that I have to search for my new record in the table

So forgive me, but this sounds like you're editing the primary key.  You say its not the same problem, but it sounds like it.
Can you pretend I'm five years old, and you're explaining this to me like I'm an idiot.  
Can you give me steps to reproduce and exactly what you're doing, preferably with screenshots.  I know I sound like I'm just wasting your time, but I'm really not.  I think this is a quirk/bug with the data entry, and to know what you're doing is useful.

Don't forget that you don't need to reorder it and put things in 'natural order'.  There is no 'natural order'.  I know the data entry looks like a spreadsheet, but think of the data as a bucket of data.  Its not sorted until you start pulling the data out to view.  Its a magic bucket.  Peering into the bucket its a jumble, but if you want the lowest number, simply reaching into the bucket and grabbing one gives you the lowest number.  The highest?  Just grab.

Anyway.  We can cross that bridge later.

So, here's my example using your database.

![image](https://user-images.githubusercontent.com/7872218/209296091-4db0f335-5fb5-44d6-a5b1-fa6002c730c4.png)

I click the new record button (1) and get a new record (2)


![image](https://user-images.githubusercontent.com/7872218/209296212-91e32648-72f4-4719-ae63-1770bc222cae.png)

I edit the details


![image](https://user-images.githubusercontent.com/7872218/209296277-4c3f4395-8b5c-40a9-8347-7fd736eb03cc.png)

I refresh (click (1) )


Values are empty

![image](https://user-images.githubusercontent.com/7872218/209296359-540daa2a-8546-42bf-b35b-53ea108a0395.png)


This is due to editing the primary key as per issue #3240.

The workaround is to edit the primary key last.  So if you want values 60, 70 and 80, enter the records first.

![image](https://user-images.githubusercontent.com/7872218/209296754-34bdfafe-3b9f-4b1b-a4b1-fdb9d2faf57a.png)

So now I can edit the primary keys

![image](https://user-images.githubusercontent.com/7872218/209296873-acf904cc-a497-400a-ac27-6fb78ee72795.png)

Clicking refresh (1) now displays full records.


![image](https://user-images.githubusercontent.com/7872218/209296985-4cce7173-7125-4c51-a75c-0faf3d46b58f.png)

Clicking on a column header to resort (1) does not lose any data.


![image](https://user-images.githubusercontent.com/7872218/209297149-d205e26c-7fd2-4636-9917-81ce23d9a6e7.png)

Viewing the log is also handy.  Ensure you've the log dock open by clicking 'View' (1) and 'SQL Log' (2)


![image](https://user-images.githubusercontent.com/7872218/209297389-030d45ba-abe7-4ab1-aa86-f561ad671eee.png)

The click the SQL log dock  tab(1) and ensure you've viewing the application log by changing the dropdown (2) to 'Application'.



If you're following a different process, please let me know and I'll do all I can to help out. :)
Thanks for your patience.  Sorry you've had this issues, and sorry you feel you can't trust this database.  We've been here a long time, still in active development and want your experience (obviously!) to be a good one.  Give us another chance and hopefully you'll be using DB4S in anger, rather than being in anger at using it! ;)

**Comment by chrisjlocke** [comment]
Also note that the primary key is usually in 'database land' an internal reference only.  Something the system uses.  Its human-readable and editable, but really it belongs to the database - it shouldn't be edited or referred to.  All the data you put in is yours, and you own that part of it.
So if you had a table of names, a record would have an ID field, but also an employee ID, or name ID, or code, etc.  Don't use the row ID to link off to a permissions table for example.
There are reasons, but the main one is rebuilding the database later.  
Some people then think 'oh, if I don't need it, I'll turn it off' and don't use row numbers at all.  Again, thats wrong.  I'll show an example later of the mess you can get into if you don't do that...

Each table should have an ID field, but this should be for DB4S's use, not yours.
There are people who argue against this - same as people argue Windows is great and wireless mice are the best, or electric cars are great.  It's purely a recommendation, but one that will keep you sane... ;)
The 'edit the primary key' can smack you in the face when you least expect it. :(

**Comment by bernard01** [comment]
I think you have a point. In my case, the primary key for a new record is always 1 not null because new entries have very high numbers like 1699070215 which I consider fresh input and not primary key change. So I overwrite that number 1.

But it is not a primary key change because the record was not saved with that number.
The table definition is such as I don't want automatic key generation. The primary key is from another application not my own so it is important.

**Comment by chrisjlocke** [comment]
If you can provide a screenshot of the data as you click 'new record', then another of the amended data, then a copy of the 'create table' command for the table.  Would be useful.

**Comment by bernard01** [comment]
New record:
![NewRecord](https://user-images.githubusercontent.com/1733616/209380868-4e084e05-6d9a-429d-b5a9-ebdb8b77ded3.png)
Data entered:
![DataEntered](https://user-images.githubusercontent.com/1733616/209382145-2b1eb03d-f86a-4fcc-b5fb-2b645e8b2152.png)
Data saved and clicked on id header for sort:
![AfterSaveWithSort](https://user-images.githubusercontent.com/1733616/209382542-31573b1b-a1af-43fb-ba2f-b4d37aca193b.png)

DDL:

CREATE TABLE "user" ( "id" INTEGER NOT NULL, "name1" TEXT NOT NULL, "name2" TEXT, "name3" TEXT, "name4" TEXT, PRIMARY KEY("id") )

You will find that in the attached zip file as well.

I noticed this setting but I never touched this:
![EditPragmas](https://user-images.githubusercontent.com/1733616/209383947-914d279f-43c5-473e-8415-5177671893ca.png)

**Comment by chrisjlocke** [comment]
Yes, so you're editing the primary key, so as per my message where you then bonked me on the head, is still valid.  You can do what you want to do, just edit the primary key last.

The 'automatic index' setting is slightly different.  This is where sqlite creates an index automatically if it thinks it'll be quicker to do so rather than hunting for data without it.  It has no bearing on what you're doing.

**Comment by bernard01** [comment]
Editing the primary key.

It's ambiguous. And I had to keep that controversy not to bang you on the head as you say.

Because I do not see this as a personal issue. And because I have to maintain that stance because this is a direction from where people using the software are coming from.

Because editing the primary key in database land is something different from the editing operation technically in GUI land. In this case it is implicit because nobody cares about that bogus initial primary key value when entering a new record.

Because it shouldn't be there. It should be null.

So this is a different bug because it is a different use case.

**Comment by bernard01** [comment]
In other words: The software as it is effectively forces people to make primary key changes even if they would never do this because the software jumps ahead and creates a bogus record with that GUI generated key which may be convenient in other cases but not in this case when the primary key is user-defined to begin with.

**Comment by bernard01** [comment]
The reason why the primary key has to be entered first is extremely important. I will explain it here.

When entering records from a foreign database source which is very common today, one has to first ascertain that this record does not already exist. So the work flow must be as follows:

1) Enter the primary key as a filter in the table view.
2) Check if a record shows up. If it shows up then review / update / do nothing
3) Enter the primary key to avoid future errors because it has just been reserved by the visual check
4) Complete data entry and save the record.

You are proposing a work flow which goes directly against this, risking duplicate useless data entry which has to be abandoned late.

I am writing this to make the use case a little clearer. A very common use case.

**Comment by chrisjlocke** [comment]
> because the software jumps ahead and creates a bogus record with that GUI generated key which may be convenient in other cases but not in this case when the primary key is user-defined to begin with

If you click the button to enter a new record, then the software is going to enter a new record.  That record has to follow the rules of the database.  The database says that field has to be a number (integer) and it cannot be null.  What do you suggest it puts in there?!

If you want to enter your own values, use the 'Add new record' dialog.

![image](https://user-images.githubusercontent.com/7872218/209426188-61e14527-2f2b-4da9-bd5c-52e8d80aeeeb.png)

Click and hold the 'Add record' dialog (1) until the dropdown appears.  Click 'Insert values'

![image](https://user-images.githubusercontent.com/7872218/209426211-005d4ce9-09c6-492e-abf0-c839f870deba.png)

Enter the values in the dialog as required.

**Comment by bernard01** [comment]
So you asking me the question what it should put into that new record.

My simple answer is "nothing". Because it usually does that for all fresh values other than the primary key.

And it should prompt me with an error message if I violate the constraints on save - not before that.

Because it is not the job of the application to enter values on my behalf for an empty new row. That's my job.

 'Insert values' via SQL is not a solution - it is just sidestepping the problem.

**Comment by chrisjlocke** [comment]
> My simple answer is "nothing"

That would be fantastic.  However, the field is specified with a 'not null' constraint.  I therefore respectfully refer you to my earlier point.

> And it should prompt me with an error message if I violate the constraints on save

It is.  It does.  It puts in a value of 1.  Try setting it to null.  It will moan when you move out of that cell.

> Because it is not the job of the application to enter values on my behalf for an empty new row. 

If the cell is set to 'integer not null' then yes, it is - that's sqlite designated primary key.  

If you don't want the cell set up with a sequential number (an alias to the rowid), then make the cell a plain integer, and also unique if you want - but remove the primary key.  However, db4s will still set that cell to 0 if you set it to 'not null' - so you'll still have to change it - accomplishing nothing.

All this is based around sqlite and how databases work.  This is a 'limitation' (loosely) on how you've defined the table.
Remove those limitations, and it'll work (loosely) how you want.

> We are pretty close I think.

I've already demonstrated and explained that there is a bug where the grid is amending the row based on what it believes is the primary key, not on what the actual primary key is.  Unfortunately, I'm not familiar with that code in question, so don't know if it's a simple fix where it can just re-read the datarow, or whether it's more convoluted and requires a complete rewrite of the internal workings.  However, I've also shown an easy workaround .. actually two.

Thanks for reporting the issue though, and hopefully we can work on a fix. :)

**Comment by bernard01** [comment]
I have a few databases with text type primary key. Never seen the problem there.

**Comment by chrisjlocke** [comment]
It's not just a primary key that's the issue.  The issue is that an integer primary key is an alias to the rowId.  This is where the number derives from - sqlite doesn't just pluck a number from the air.
You can prove this quite easily.  Create a table with just two fields.  Add five records.  You'll have five records of nulls.  Add a new column of an integer, and set it as the primary key.  SQLite (not DB4S) fills in the values of those records with the rowId.

![image](https://user-images.githubusercontent.com/7872218/209682622-f8764d0b-94c5-41d4-a884-6815f3e60017.png)

And here's the table with the rowId.

![image](https://user-images.githubusercontent.com/7872218/209682762-70e440af-5b7b-48c4-afe5-fe50e261f842.png)

> I have a few databases with text type primary key. Never seen the problem there.

This is due to DB4S using the rowId as the row identifier to edit rows.  This is shown in the log.

![image](https://user-images.githubusercontent.com/7872218/209683146-a67faeeb-41af-4a97-86d4-1ad957bad8db.png)

So editing cell values has no effect on the rowId.  However, for an integer primary key field, as this is an alias to the rowId, changing this row also changes the rowId. It's this 'bug' in the original post. :)

**Comment by chrisjlocke** [comment]
Just noticed the screenshot of the log shows _rowid_ and I wobble on about rowId.  SQLite has three aliases.  
"The [rowid](https://www.sqlite.org/lang_createtable.html#rowid) of a [rowid](https://www.sqlite.org/lang_createtable.html#rowid) table can be accessed (or changed) by reading or writing to any of the "rowid" or "oid" or "_rowid_" columns. Except, if there is a declared columns in the table that use those special names, then those names refer to the declared columns, not to the underlying rowid."
https://www.sqlite.org/rowidtable.html

**Comment by i-s-o** [comment]
Thanks to this thread, I learned a few new tricks (I didn't know about *click-and-hold* action on toolbar buttons). Since it took me some time to digest what's going on, I thought I'd summarize this issue for my future self, and maybe get a confirmation from the devs whether I got it right:

  1. The UI labels has specific, strict meanings, i.e.
     * Clicking on *New Record* literally results in "adding a new record to the table and committing it with default field values if they satisfy the table constraints". As such, it doesn't let the user *preview/edit* the values before they are added. What the user sees in the browser grid is already committed to the database, and whatever edits follow are simply `UPDATE ...` actions.
     * If the user needs to enter specific values before a new record is inserted to the table, the right choice is *Insert Values ...* option, which will open a dialog for the user to enter those specific values.
  2. OP's experience is still less than ideal, IMO this is still a UI bug.
     * Good news: it looks like this behavior was fixed on the master branch in early June as a side effect of fixing another issue: #3026. Now, we just need to have a new release so that this fix trickles to all platforms automatically ;-).
     * Starting with the 2022-06-11 nightly, if the table has a PK without autoincrement (as in OP's test database above), clicking on *New Record* actually invokes *Insert Values ...* dialog, and the user is asked to provide the value for the PK.

**Comment by chrisjlocke** [comment]
> What the user sees in the browser grid is already committed to the database

This is 92.1% correct.  It's committed to the journal, and when you click 'Write changes', its actually committed to the bits and bytes of the database - you can click 'Discard charges' to ... well, discard.

#3026 is a slightly different issue, and not related this this one.
However, for the avoidance of doubt, this issue reported here is still a bug.  After editing the primary key, DB4S should take note and write it down, not just pin it on the board and forget all about it.  Oopsie.
So unfortunately, the fix you mention (and thanks for pointing it out) is not going to help.  Also, a new release isn't necessary - we produce nightly builds, and on Linux, you can easily build DB4S yourself.

**Comment by i-s-o** [comment]
Journal v. committing to the actual database: got it, 100% agree on that.

I also agree with the bug you described, which is #3240.

However, in subsequent comments, the OP [indicated that his issue was not the same](https://github.com/sqlitebrowser/sqlitebrowser/issues/3248#issuecomment-1363400703), and then [said](https://github.com/sqlitebrowser/sqlitebrowser/issues/3248#issuecomment-1364621725): 

> And it should prompt me with an error message if I violate the constraints on save - not before that.

My earlier comment was simply to point out that with a recent nightly, I don't run into the issue raised by OP because clicking on "New Record" button bypasses the buggy behavior:

https://user-images.githubusercontent.com/1661549/210101572-dec4a14f-b866-4971-964d-3812738ab28d.mp4

**Comment by chrisjlocke** [comment]
> However, in subsequent comments, the OP indicated that his issue was not the same, 

And in subsequent comments, I pointed out it was.  The OP was editing the primary key, which is the same issue.  Editing the primary key results in the same scenario - I even provided screenshots.

> because clicking on "New Record" button bypasses the buggy behavior

Not really.  Well, not exactly.  It appears you have some index on name2?  A unique index I assume?  Or not null?  I don't know the exact rules, but it appears you can have one null value.  When you add the second record, DB4S can't add a second null record, so therefore has to display the dialog.

**Comment by bernard01** [comment]
I can't really see what the discussion has achieved.

I downloaded the latest release candidate.

Created a new database and a table in it with an integer primary key (NOT auto-increment) and two more fields..

I create a new record. The first thing is that the browser actually generates a primary key with value 1. But my primary key is NOT 1 but some other value such as 5.

So the browser mucks around with my data against my intention and puts something into the primary key as if it was auto-increment which it isn't which is wrong. I change the key value to 5 as intended and enter values in the other fields and save the record.

When I look at it later, then the data is lost.

I have never seen anything as stupid as this in a scenario as basic as this.

**Comment by i-s-o** [comment]
I agree that this is a nasty bug, and that you'd need to know a ton about how [SQLite3 ](https://www.sqlite.org/withoutrowid.html) and [DB4S](https://github.com/sqlitebrowser/sqlitebrowser/issues/3248#issuecomment-1365956648) works to avoid it!

TL;DR:
-------
Don't use "INTEGER PRIMARY KEY" when you declare your primary key.
Instead, use "INT PRIMARY KEY".

Explanation:
-------------
As explained in earlier posts, "INTEGER PRIMARY KEY" is a special case when declaring a variable: with this key phrase, SQLite3 creates an alias to ROWID instead of actually creating a new variable. Also note that ROWID always AUTO INCREMENTs. Unfortunately, the mechanism used by the Browser pane in DB4S relies on ROWID, and if you change the value in your primary key (i.e. ROWID), you run into this bug.
```sql
-- BAD: avoid changing the PK auto-incremented value show for new records in the Browser panel
CREATE TABLE "integer_pk" (
  "id" INTEGER PRIMARY KEY,
  "name1" TEXT NOT NULL,
  "name2" TEXT,
);
```

Workarounds:
---------------
1. Declare your primary key as INT PRIMARY KEY: with this declaration, SQLite3 will actually create a new variable independent of ROWID. Then, the DB4S's Browser panel will handle new records properly.
```sql
-- Simple solution (but quite obscure)
CREATE TABLE "int_pk_with_rowid" (
  "id" INT PRIMARY KEY,
  "name1" TEXT NOT NULL,
  "name2" TEXT,
);
```

2. Declare your table WITHOUT ROWID: in absence of the ROWID column, since DB4S's Browser panel cant' insert a new [blank] row, DB4S will display the "Add New Record" dialog, which will let you enter the value you want for the primary key.
```sql
-- Alternative solution: this will force "Add New Record" dialog
CREATE TABLE "integer_pk_without_rowid" (
  "id" INTEGER PRIMARY KEY,
  "name1" TEXT NOT NULL,
  "name2" TEXT,
) WITHOUT ROWID;
```

Possible improvement in DB4S:
----------------------------------
DB4S should mark the column declared as INTEGER PRIMARY KEY as *read only* to avoid this problem. This will force users to choose one of the workarounds listed above.

**Comment by bernard01** [comment]
Thank you guys for all your efforts. I applied the INT workaround and will work with it from now on.

After all, the browser is a great product and that it is why it was worth the pain.

**Comment by chrisjlocke** [comment]
> So the browser mucks around with my data against my intention and puts something into the primary key as if it was auto-increment which it isn't which is wrong.

For the avoidance of doubt, and to confirm your point, this is **not** the 'browser' (DB4S) 'mucking about' with your data.
If you set a field as primary key and its an integer, then that field is also an alias for the rowId for that table, which is an auto-incrementing field.  

As explained previously, you can edit this value, but the edits have to be done in steps.  If you perform step 1 first, then you can perform step 2.

> I have never seen anything as stupid as this in a scenario as basic as this.

This is like having a front door with a big number 5 on it.  Someone says to you, 'go up to door 5 and paint a number 7 on it.'  Your next instruction is 'go to door 5 and add a letterbox to it.' as at the time that was written, door 5 was door 5, not door 7.  OK, you're clever enough to know 'hey, I just painted door 5 to door 7 .. i wonder if that's the same door?' but DB4S errs on the side of caution, so bails going 'i don't know where door 5 is'.

Editing primary keys isn't a 'basic scenario'.  You don't normally do this,
The usual rule of thumb when creating tables is that you add a primary key, then all the fields you need.
This way, there isn't a need to edit the primary key.  The primary key shouldn't be 'user data'.

> I can't really see what the discussion has achieved.

After all my efforts, I'm so glad to have helped.  Now I feel worthless. :(
